genrl/deep/agents/dqn/base.py

```diff
@@ -0,0 +1,246 @@
+            exploration with time)
+        seed (int): Seed for randomness
+        render (bool): Should the env be rendered during training?
+        device (str): Hardware being used for training. Options:
+            ["cuda" -> GPU, "cpu" -> CPU]
+    """
+
+    def __init__(
+        self,
+        *args,
+        max_epsilon: float = 1.0,
+        min_epsilon: float = 0.01,
+        epsilon_decay: int = 1000,
+        **kwargs
+    ):
+        super(BaseDQN, self).__init__(*args, **kwargs)
+        self.max_epsilon = max_epsilon
+        self.min_epsilon = min_epsilon
+        self.epsilon_decay = epsilon_decay
+
+    def create_model(self, *args) -> None:
+        """Function to initialize Q-value model
+
+        This will create the Q-value function of the agent. Depends on the network type,
+        types of Q-value models needed.
+
+        Supported:
+            "v" and "Qs" to `get_value_from_name` -> Regular Q-value function
+            "dv" and "mlpdueling" -> Dueling Q-value function
+            "dv" and "mlpnoisy" -> Noisy Q-value function
+            "dv" and "mlpcategorical -> Categorical Q-value function
+
+        You can replace "mlp" with "cnn" for using a CNN based Q-value function
+        """
+        input_dim, action_dim, _, _ = get_env_properties(self.env, self.network_type)
+
+        self.model = get_model("v", self.network_type)(
+            input_dim, action_dim, "Qs", self.layers
+        )
+        self.target_model = deepcopy(self.model)
+
+        self.replay_buffer = self.buffer_class(self.replay_size, *args)
+        self.optimizer = opt.Adam(self.model.parameters(), lr=self.lr_value)
+
+    def update_target_model(self) -> None:
+        """Function to update the target Q model
+
+        Updates the target model with the training model's weights when called
+        """
+        self.target_model.load_state_dict(self.model.state_dict())
+
+    def update_params_before_select_action(self, timestep: int) -> None:
+        """Update necessary parameters before selecting an action
+
+        This updates the epsilon (exploration rate) of the agent every timestep
+
+        Args:
+            timestep (int): Timestep of training
+        """
+        self.timestep = timestep
+        self.epsilon = self.calculate_epsilon_by_frame()
+        self.logs["epsilon"].append(self.epsilon)
+
+    def select_action(
+        self, state: np.ndarray, deterministic: bool = False
+    ) -> np.ndarray:
+        """Select action given state
+
+        Epsilon-greedy action-selection
+
+        Args:
+            state (:obj:`np.ndarray`): Current state of the environment
+            deterministic (bool): Should the policy be deterministic or stochastic
+
+        Returns:
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
```

Comment:
`nit`
Returning inside if statements is not preferable generally as a programming practice.

Reply:
Ok, will change this